File: src/main/java/beam/agentsim/events/ReplanningEvent.java
package beam.agentsim.events;

import org.matsim.api.core.v01.Id;
import org.matsim.api.core.v01.events.Event;
import org.matsim.api.core.v01.population.Person;
import org.matsim.core.api.internal.HasPersonId;

import java.util.Map;

public class ReplanningEvent extends Event implements HasPersonId {

    public final static String EVENT_TYPE = "Replanning";
    public final static String ATTRIBUTE_PERSON = "person";
    public final static String ATTRIBUTE_REPLANNING_REASON = "reason";
    public final static String ATTRIBUTE_START_COORDINATE_X = "startX";
    public final static String ATTRIBUTE_START_COORDINATE_Y = "startY";
    public final static String ATTRIBUTE_END_COORDINATE_X = "endX";
    public final static String ATTRIBUTE_END_COORDINATE_Y = "endY";

    private final Id<Person> personId;
    private final String reason;

    private final String startX;

    private final String startY;

    private final String endX;

    private final String endY;

    public ReplanningEvent(final double time, final Id<Person> personId, final String reason) {
        super(time);
        this.personId = personId;
        this.reason = reason;
        this.startX = "";
        this.startY = "";
        this.endX = "";
        this.endY = "";
    }

    public ReplanningEvent(final double time,
                           final Id<Person> personId,
                           final String reason,
                           final Double startX,
                           final Double startY,
                           final Double endX,
                           final Double endY) {
        super(time);
        this.personId = personId;
        this.reason = reason;
        this.startX = startX.toString();
        this.startY = startY.toString();
        this.endX = endX.toString();
        this.endY = endY.toString();
    }

    public ReplanningEvent(final double time,
                           final Id<Person> personId,
                           final String reason,
                           final Double startX,
                           final Double startY) {
        super(time);
        this.personId = personId;
        this.reason = reason;
        this.startX = startX.toString();
        this.startY = startY.toString();
        this.endX = "";
        this.endY = "";
    }

    public static ReplanningEvent apply(Event event) {
        if (!(event instanceof ReplanningEvent) && EVENT_TYPE.equalsIgnoreCase(event.getEventType())) {
            Map<String, String> attr = event.getAttributes();
            return new ReplanningEvent(event.getTime(),
                    Id.createPersonId(attr.get(ATTRIBUTE_PERSON)),
                    attr.get(ATTRIBUTE_REPLANNING_REASON),
                    Double.valueOf(attr.get(ATTRIBUTE_START_COORDINATE_X).replaceFirst("^$", "0.0")),
                    Double.valueOf(attr.get(ATTRIBUTE_START_COORDINATE_Y).replaceFirst("^$", "0.0")),
                    Double.valueOf(attr.get(ATTRIBUTE_END_COORDINATE_X).replaceFirst("^$", "0.0")),
                    Double.valueOf(attr.get(ATTRIBUTE_END_COORDINATE_Y).replaceFirst("^$", "0.0"))
            );
        }
        return (ReplanningEvent) event;
    }

    @Override
    public Id<Person> getPersonId() {
        return personId;
    }

    @Override
    public String getEventType() {
        return EVENT_TYPE;
    }

    public String getReason(){ return reason; }

    @Override
    public Map<String, String> getAttributes() {
        Map<String, String> attr = super.getAttributes();
        attr.put(ATTRIBUTE_PERSON, personId.toString());
        attr.put(ATTRIBUTE_REPLANNING_REASON, reason);
        attr.put(ATTRIBUTE_START_COORDINATE_X, startX);
        attr.put(ATTRIBUTE_START_COORDINATE_Y, startY);
        attr.put(ATTRIBUTE_END_COORDINATE_X, endX);
        attr.put(ATTRIBUTE_END_COORDINATE_Y, endY);
        return attr;
    }
}

File: src/main/scala/beam/agentsim/events/ReplanningEvent.scala
package beam.agentsim.events

import beam.agentsim.events.ReplanningEvent._
import beam.router.Modes.BeamMode
import org.matsim.api.core.v01.events.Event
import org.matsim.api.core.v01.population.Person
import org.matsim.api.core.v01.{Coord, Id}
import org.matsim.vehicles.Vehicle

import java.util
import scala.jdk.CollectionConverters.mapAsScalaMapConverter

/**
  * Event capturing the details of a trip replanning at runtime.
  */
case class ReplanningEvent(
  tick: Double,
  personId: Id[Person],
  reason: String,
  vehicleId: Option[Id[Vehicle]] = None,
  legMode: Option[BeamMode] = None,
  start: Coord,
  end: Option[Coord] = None
) extends Event(tick)
    with ScalaEvent {

  override def getEventType: String = EVENT_TYPE

  override def getAttributes: util.Map[String, String] = {
    val attr = super.getAttributes
    attr.put(ATTRIBUTE_PERSON, personId.toString)
    attr.put(ATTRIBUTE_VEHICLE, vehicleId.fold("")(_.toString))
    attr.put(ATTRIBUTE_LEG_MODE, legMode.fold("")(_.value))
    attr.put(ATTRIBUTE_REPLANNING_REASON, reason)
    attr.put(ATTRIBUTE_START_COORDINATE_X, start.getX.toString)
    attr.put(ATTRIBUTE_START_COORDINATE_Y, start.getY.toString)
    attr.put(ATTRIBUTE_END_COORDINATE_X, end.fold("")(_.getX.toString))
    attr.put(ATTRIBUTE_END_COORDINATE_Y, end.fold("")(_.getY.toString))
    attr
  }
}

object ReplanningEvent {
  val EVENT_TYPE = "Replanning"
  val ATTRIBUTE_PERSON = "person"
  val ATTRIBUTE_VEHICLE = "vehicle"
  val ATTRIBUTE_REPLANNING_REASON = "reason"
  val ATTRIBUTE_LEG_MODE = "mode"
  val ATTRIBUTE_START_COORDINATE_X = "startX"
  val ATTRIBUTE_START_COORDINATE_Y = "startY"
  val ATTRIBUTE_END_COORDINATE_X = "endX"
  val ATTRIBUTE_END_COORDINATE_Y = "endY"

  def apply(genericEvent: Event): ReplanningEvent = {
    assert(genericEvent.getEventType == EVENT_TYPE)
    val attr = genericEvent.getAttributes.asScala
    val time: Double = genericEvent.getTime
    val personId: Id[Person] = Id.createPersonId(attr(ATTRIBUTE_PERSON))
    val reason: String = attr.getOrElse(ATTRIBUTE_REPLANNING_REASON, "")
    val vehicleId: Option[Id[Vehicle]] = attr.get(ATTRIBUTE_VEHICLE).map(Id.createVehicleId)
    val legMode: Option[BeamMode] = BeamMode.fromString(attr.getOrElse(ATTRIBUTE_LEG_MODE, ""))
    val startX: Double = attr.getOrElse(ATTRIBUTE_START_COORDINATE_X, "0.0").replaceFirst("^$", "0.0").toDouble
    val startY: Double = attr.getOrElse(ATTRIBUTE_START_COORDINATE_Y, "0.0").replaceFirst("^$", "0.0").toDouble
    val endX: Option[Double] = attr.get(ATTRIBUTE_END_COORDINATE_X).map(_.replaceFirst("^$", "0.0").toDouble)
    val endY: Option[Double] = attr.get(ATTRIBUTE_END_COORDINATE_Y).map(_.replaceFirst("^$", "0.0").toDouble)
    ReplanningEvent(
      time,
      personId,
      reason,
      vehicleId,
      legMode,
      new Coord(startX, startY),
      for {
        x <- endX
        y <- endY
      } yield new Coord(x, y)
    )
  }
}
